test/exe/main_common_test.cc

```diff
@@ -40,7 +40,20 @@ class MainCommonTest : public testing::TestWithParam<Network::Address::IpVersion
             TestEnvironment::PortMap(), GetParam())),
         random_string_(fmt::format("{}", computeBaseId())),
         argv_({"envoy-static", "--base-id", random_string_.c_str(), "-c", config_file_.c_str(),
-               nullptr}) {}
+               nullptr}) {
+    // The test main() sets the ThreadFactorySingleton since it is required by all other tests not
+    // instantiating their own MainCommon.
+    // Reset the singleton to a nullptr to avoid triggering an assertion when MainCommonBase() calls
+    // set() in the tests below.
+    Thread::ThreadFactorySingleton::set(nullptr);
+  }
+
+  ~MainCommonTest() override {
+    // This is ugly, but necessary to enable a stronger ASSERT() in ThreadFactorySingleton::set().
+    // The singleton needs to be reset to a non nullptr value such that when the constructor runs
+    // again, the ThreadFactorySingleton::set(nullptr) does not trigger the assertion.
+    Thread::ThreadFactorySingleton::set(&Thread::threadFactoryForTest());
```

Comment:
this is getting a little complex (IMO life often gets this way with singletons but that ship has sailed)...so now here's what I hope is a more helpful suggestion.

Can we make the setting of the singleton the responsibility of main() ?  No one else should do it. We would not do it in main_common.cc (or have to undo it in main_common_test.cc). We have test/main.cc and source/exe/main.cc which would both need to set it.

In Google we have a private main() and a private ThreadFactory, and when we import your change we'd just have to make a modification our google-specific main()s. @PiotrSikora will be on import next week so he should be aware of this suggestion and what you decide to do with it :)

Reply:
per f2f discussion I agree this is the cleanest method, and is well commented.

Reply:
To provide some context into the resolution to the issue raised by @jmarantz, having the MainCommonBase constructor set the singleton as is currently done allows other main()s to _not_ have to change since the platform specific ThreadFactory is already passed as an argument to the aforementioned constructor.

Reply:
:( but ok.